android/src/main/kotlin/co/doneservices/callkeep/CallKeepNotificationManager.kt

```diff
@@ -0,0 +1,439 @@
+        )
+    }
+
+    private fun getTimeOutPendingIntent(id: Int, data: Bundle): PendingIntent {
+        val timeOutIntent = CallKeepBroadcastReceiver.getIntentTimeout(context, data)
+        return PendingIntent.getBroadcast(
+                context,
+                id,
+                timeOutIntent,
+                getFlagPendingIntent()
+        )
+    }
+
+    private fun getCallbackPendingIntent(id: Int, data: Bundle): PendingIntent {
+        val intent = context.packageManager.getLaunchIntentForPackage(context.packageName)?.cloneFilter()
+        intent?.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+        if (intent != null) {
+            val intentTransparent = TransparentActivity.getIntentCallback(context, data)
+            return PendingIntent.getActivities(
+                    context,
+                    id,
+                    arrayOf(intent, intentTransparent),
+                    getFlagPendingIntent()
+            )
+        } else {
+            val acceptIntent = CallKeepBroadcastReceiver.getIntentCallback(context, data)
+            return PendingIntent.getBroadcast(
+                    context,
+                    id,
+                    acceptIntent,
+                    getFlagPendingIntent()
+            )
+        }
+    }
+
+    private fun getActivityPendingIntent(id: Int, data: Bundle): PendingIntent {
+        val intent = IncomingCallActivity.getIntent(context,data)
+        return PendingIntent.getActivity(context, id, intent, getFlagPendingIntent())
+    }
+
+    private fun getAppPendingIntent(id: Int, data: Bundle): PendingIntent {
+        val intent: Intent? = context.packageManager.getLaunchIntentForPackage(context.packageName)
+        intent?.putExtra(CallKeepBroadcastReceiver.EXTRA_CALLKEEP_INCOMING_DATA, data)
+        return PendingIntent.getActivity(context, id, intent, getFlagPendingIntent())
+    }
+
+    private fun getFlagPendingIntent(): Int {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
+        } else {
+            PendingIntent.FLAG_UPDATE_CURRENT
+        }
+    }
+
+    private fun getNotificationManager(): NotificationManagerCompat {
+        return NotificationManagerCompat.from(context)
+    }
+
+
+    private fun getPicassoInstance(context: Context, headers: HashMap<String, Any?>): Picasso {
+        val client = OkHttpClient.Builder()
+                .addNetworkInterceptor { chain ->
+                    val newRequestBuilder: okhttp3.Request.Builder = chain.request().newBuilder()
+                    for ((key, value) in headers) {
+                        newRequestBuilder.addHeader(key, value.toString())
+                    }
+                    chain.proceed(newRequestBuilder.build())
+                }
+                .build()
+        return Picasso.Builder(context)
+                .downloader(OkHttp3Downloader(client))
+                .build()
+    }
+
+
```

Comment:
This file is sooo big better to chunk down a bit

Reply:
Agreed, but I think we can improve on it in upcoming PRs